File: global_interp/c/polint3.c
void polin3(const double *x1a, const double *x2a, const double *x3a,
                   const double *ya, double x1, double x2, double x3,
                   double *y, double *dy, int ordn)
{
    // ya is ordn x ordn x ordn in Fortran layout (column-major)
    #define YA3(i,j,k) ya[(i) + ordn*((j) + ordn*(k))]  // i,j,k: 0..ordn-1

    int j, k;
    double dy_temp;

    // yatmp(j,k) in Fortran code is ordn x ordn, treat column-major:
    // yatmp(j,k) -> yatmp[j + ordn*k]
    double *yatmp = (double*)malloc((size_t)ordn * (size_t)ordn * sizeof(double));
    double *ymtmp = (double*)malloc((size_t)ordn * sizeof(double));
    if (!yatmp || !ymtmp) {
        fprintf(stderr, "polin3: malloc failed\n");
        exit(1);
    }
    #define YAT(j,k) yatmp[(j) + ordn*(k)]

    for (k = 0; k < ordn; k++) {
        for (j = 0; j < ordn; j++) {
            // call polint(x1a, ya(:,j,k), x1, yatmp(j,k), dy_temp)
            // ya(:,j,k) contiguous: base is &YA3(0,j,k)
            polint(x1a, &YA3(0, j, k), x1, &YAT(j, k), &dy_temp, ordn);
        }
    }

    for (k = 0; k < ordn; k++) {
        // call polint(x2a, yatmp(:,k), x2, ymtmp(k), dy_temp)
        polint(x2a, &YAT(0, k), x2, &ymtmp[k], &dy_temp, ordn);
    }

    polint(x3a, ymtmp, x3, y, dy, ordn);

    #undef YAT
    free(yatmp);
    free(ymtmp);
    #undef YA3
}

File: global_interp/fortran/polint3.f90
!------------------------------------------------------------------------------
!
! interpolation in 3 dimensions, follow zyx order
!
!------------------------------------------------------------------------------
  subroutine polin3(x1a,x2a,x3a,ya,x1,x2,x3,y,dy,ordn)
  implicit none

  integer,intent(in) :: ordn
  real*8, dimension(1:ordn), intent(in) :: x1a,x2a,x3a
  real*8, dimension(1:ordn,1:ordn,1:ordn), intent(in) :: ya
  real*8, intent(in) :: x1,x2,x3
  real*8, intent(out) :: y,dy

  integer  :: j, k
  real*8, dimension(ordn,ordn) :: yatmp
  real*8, dimension(ordn) :: ymtmp
  real*8 :: dy_temp

  do k=1,ordn
    do j=1,ordn
      call polint(x1a, ya(:,j,k), x1, yatmp(j,k), dy_temp, ordn)
    end do
  end do
  do k=1,ordn
    call polint(x2a, yatmp(:,k), x2, ymtmp(k), dy_temp, ordn)
  end do
  call polint(x3a, ymtmp, x3, y, dy, ordn)


  return
  end subroutine polin3